### PR TITLE
SALTO-5533 - Salesforce: Fix Quickfetch for allowReferenceTo records

### PR DESCRIPTION
**Pull request by tomsellek**
When fetch-with-changes-detection is enabled, we only attempt to fetch instances that are referred to from instances that changed.
Instead, we should fetch instances that are referenced by instances in the workspace too, in case any of the referred-to instances changed.

---

~We cheat a little by treating `InstanceElement.value` as a `SalesforceRecord` for the purpose of finding references. This is because the data in the workspace/`elementsSource` is stored as `InstanceElement`s, but the fetched records are `SalesforceRecord`s.~

There is an issue where if the user deleted the last reference to a data record, the referred record will not be removed from the workspace. This is because we don't identify the deletion of the referring record (because we don't support deletions in quickfetch)

---
_Release Notes_: 
Salesforce: Quickfetch now works for data records of types listed in `allowReferenceTo`

---
_User Notifications_: 
N/A

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 1]
```diff
@@ -536,19 +541,72 @@ export const getAllInstances = async (
...
+
+  const fetchedBaseRecordByTypeAndId = _.mapValues(
+    fetchedBaseRecordsByType,
+    (records) => _.keyBy(records, (record) => record[CUSTOM_OBJECT_ID_FIELD]),
   )
+
+  const elementSourceRecordsByTypeAndId = _.mapValues(
+    elementSourceRecordsByType,
+    (records) => _.keyBy(records, (record) => record[CUSTOM_OBJECT_ID_FIELD]),
+  )
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:589]: I would export all the logic of creating **elementSourceRecordsByTypeAndId** to a dedicated method

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts [hunk 2]
```diff
@@ -778,6 +836,80 @@ const filterCreator: RemoteFilterCreator = ({ client, config }) => ({
   name: 'customObjectsInstancesFilter',
   remote: true,
   onFetch: async (elements: Element[]): Promise<FilterResult> => {
+    const getAllCustomObjects = async (
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/custom_objects_instances.ts:839]: Is this existing code that moved?

**Review by tamtamirr — COMMENTED** [review]
LGTM! Let's test older version with this new one, and see there's no regression in full fetch. Added some comments. There's quite a diff in this PR, but overall looks good!

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts (hunk 2 above)
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:839]: Yes, from line 793 of the old file

File: packages/salesforce-adapter/src/filters/custom_objects_instances.ts (hunk 1 above)
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_objects_instances.ts:589]: Ack

**Review by tamtamirr — APPROVED** [review]
LGTM!

**Pull request merged**